File: tests/java_static/src/test/java/myairtable/TestCacheStore.java
package myairtable;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.util.concurrent.atomic.AtomicLong;
import org.junit.jupiter.api.Test;

/**
 * J-port of Kotlin K2.7 — CacheStore TTL expiry, LRU eviction, and scope invalidation. Parity:
 * Kotlin/Swift TestCacheStore. Uses an injected clock ({@link AtomicLong} behind a {@code
 * LongSupplier}) for TTL determinism.
 */
class TestCacheStore {

  private static CacheStore.Key key() {
    return key("tbl1", "k");
  }

  private static CacheStore.Key keyedOn(String keyedOn) {
    return key("tbl1", keyedOn);
  }

  private static CacheStore.Key key(String table, String keyedOn) {
    return new CacheStore.Key(table, keyedOn);
  }

  @Test
  void zeroTtlDisablesCaching() {
    CacheStore cache = new CacheStore(0.0);
    cache.set(key(), "payload");
    assertNull(cache.get(key()));
    assertEquals(0, cache.count());
  }

  @Test
  void hitWithinTtlReturnsPayload() {
    CacheStore cache = new CacheStore(60.0);
    cache.set(key(), "payload");
    assertEquals("payload", cache.get(key()));
  }

  @Test
  void ttlOnlyConstructorAppliesADefaultLruCap() {
    // JR-M5: a long-lived TTL cache must be bounded — the TTL-only ctor applies
    // DEFAULT_MAX_ENTRIES rather than growing without limit (expiry is lazy).
    CacheStore cache = new CacheStore(60.0);
    assertEquals(CacheStore.DEFAULT_MAX_ENTRIES, cache.getMaxEntries());
  }

  @Test
  void expiredEntryIsEvictedOnAccess() {
    AtomicLong now = new AtomicLong(0L);
    CacheStore cache = new CacheStore(1.0, null, now::get);
    cache.set(key(), "payload");
    now.set(2_000L);
    assertNull(cache.get(key()));
    assertEquals(0, cache.count());
  }

  @Test
  void perCallTtlOverridesDefault() {
    AtomicLong now = new AtomicLong(0L);
    CacheStore cache = new CacheStore(1.0, null, now::get);
    cache.set(key(), "payload");
    now.set(5_000L);
    assertNotNull(cache.get(key(), 60.0));
  }

  @Test
  void capEvictsExactlyDownToMaxEntries() {
    CacheStore cache = new CacheStore(60.0, 2);
    cache.set(keyedOn("a"), "1");
    cache.set(keyedOn("b"), "2");
    cache.set(keyedOn("c"), "3");
    assertEquals(2, cache.count());
    assertNull(cache.get(keyedOn("a")), "oldest entry evicted");
    assertNotNull(cache.get(keyedOn("c")));
  }

  @Test
  void getRefreshesRecencyEvictionIsLruNotFifo() {
    CacheStore cache = new CacheStore(60.0, 2);
    cache.set(keyedOn("a"), "1");
    cache.set(keyedOn("b"), "2");
    cache.get(keyedOn("a")); // refresh "a" — "b" is now least recent
    cache.set(keyedOn("c"), "3");
    assertNotNull(cache.get(keyedOn("a")));
    assertNull(cache.get(keyedOn("b")));
  }

  @Test
  void setOnAnExistingKeyRefreshesRecency() {
    CacheStore cache = new CacheStore(60.0, 2);
    cache.set(keyedOn("a"), "1");
    cache.set(keyedOn("b"), "2");
    cache.set(keyedOn("a"), "1-updated"); // refresh "a"
    cache.set(keyedOn("c"), "3");
    assertEquals("1-updated", cache.get(keyedOn("a")));
    assertNull(cache.get(keyedOn("b")));
  }

  @Test
  void tableInvalidationOnlyDropsThatTable() {
    CacheStore cache = new CacheStore(60.0);
    cache.set(key("tbl1", "k"), "1");
    cache.set(key("tbl2", "k"), "2");
    cache.invalidateTable("tbl1");
    assertNull(cache.get(key("tbl1", "k")));
    assertNotNull(cache.get(key("tbl2", "k")));
  }

  @Test
  void allInvalidationDropsEverything() {
    CacheStore cache = new CacheStore(60.0);
    cache.set(key("tbl1", "k"), "1");
    cache.set(key("tbl2", "k"), "2");
    cache.clear();
    assertEquals(0, cache.count());
  }
}

File: tests/kotlin_static/src/test/kotlin/myairtable/TestCacheStore.kt
package myairtable

import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

/**
 * K2.7 — CacheStore TTL expiry, LRU eviction, and scope invalidation.
 * Parity: Swift TestCacheStore. Uses an injected clock for TTL determinism.
 */
class TestCacheStore {
    private fun key(
        table: String = "tbl1",
        keyedOn: String = "k",
    ) = CacheStore.Key(table, keyedOn)

    @Test
    fun zeroTtlDisablesCaching() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 0.0)
            cache.set(key(), "payload")
            assertNull(cache.get(key()))
            assertEquals(0, cache.count())
        }

    @Test
    fun hitWithinTtlReturnsPayload() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 60.0)
            cache.set(key(), "payload")
            assertEquals("payload", cache.get(key()))
        }

    @Test
    fun expiredEntryIsEvictedOnAccess() =
        runTest {
            var now = 0L
            val cache = CacheStore(defaultTtlSeconds = 1.0, nowMillis = { now })
            cache.set(key(), "payload")
            now = 2_000L
            assertNull(cache.get(key()))
            assertEquals(0, cache.count())
        }

    @Test
    fun perCallTtlOverridesDefault() =
        runTest {
            var now = 0L
            val cache = CacheStore(defaultTtlSeconds = 1.0, nowMillis = { now })
            cache.set(key(), "payload")
            now = 5_000L
            assertNotNull(cache.get(key(), ttlSeconds = 60.0))
        }

    @Test
    fun capEvictsExactlyDownToMaxEntries() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 60.0, maxEntries = 2)
            cache.set(key(keyedOn = "a"), "1")
            cache.set(key(keyedOn = "b"), "2")
            cache.set(key(keyedOn = "c"), "3")
            assertEquals(2, cache.count())
            assertNull(cache.get(key(keyedOn = "a")), "oldest entry evicted")
            assertNotNull(cache.get(key(keyedOn = "c")))
        }

    @Test
    fun getRefreshesRecencyEvictionIsLruNotFifo() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 60.0, maxEntries = 2)
            cache.set(key(keyedOn = "a"), "1")
            cache.set(key(keyedOn = "b"), "2")
            cache.get(key(keyedOn = "a")) // refresh "a" — "b" is now least recent
            cache.set(key(keyedOn = "c"), "3")
            assertNotNull(cache.get(key(keyedOn = "a")))
            assertNull(cache.get(key(keyedOn = "b")))
        }

    @Test
    fun setOnAnExistingKeyRefreshesRecency() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 60.0, maxEntries = 2)
            cache.set(key(keyedOn = "a"), "1")
            cache.set(key(keyedOn = "b"), "2")
            cache.set(key(keyedOn = "a"), "1-updated") // refresh "a"
            cache.set(key(keyedOn = "c"), "3")
            assertEquals("1-updated", cache.get(key(keyedOn = "a")))
            assertNull(cache.get(key(keyedOn = "b")))
        }

    @Test
    fun tableInvalidationOnlyDropsThatTable() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 60.0)
            cache.set(key(table = "tbl1"), "1")
            cache.set(key(table = "tbl2"), "2")
            cache.invalidate(CacheStore.Scope.Table("tbl1"))
            assertNull(cache.get(key(table = "tbl1")))
            assertNotNull(cache.get(key(table = "tbl2")))
        }

    @Test
    fun allInvalidationDropsEverything() =
        runTest {
            val cache = CacheStore(defaultTtlSeconds = 60.0)
            cache.set(key(table = "tbl1"), "1")
            cache.set(key(table = "tbl2"), "2")
            cache.clear()
            assertEquals(0, cache.count())
        }
}
